File: OOP/oop-labs/Lab2/DateUtil.cpp
//
// Created by Bing Chen on 2018/5/6.
//

#include "DateUtil.h"

bool isLeapYear(int year) {
    return year % 400 == 0 || (year % 100 != 0 && year % 4 == 0);
}

int getTotalDayOfMonth(int year, int month) {
    if (month == 4 || month == 6 || month == 9 || month == 11) {
        return 30;
    } else if (isLeapYear(year) && month == 2) {
        return 29;
    } else if (month == 2) {
        return 28;
    } else {
        return 31;
    }
}

bool isBefore(Date &a, Date &b) {
    if (a.getYear() > b.getYear()) {
        return false;
    } else if (a.getYear() < b.getYear()) {
        return true;
    } else if (a.getMonth() > b.getMonth()) {
        return false;
    } else if (a.getMonth() < b.getMonth()) {
        return true;
    } else {
        return a.getDay() <= b.getDay();
    }
}

bool isTheSameDay(Date &a, Date &b) {
    return a.getDay() == b.getDay() && a.getMonth() == b.getMonth() && a.getYear() == b.getYear();
}

void dateIncrease(Date &date) {
    int year = date.getYear();
    int month = date.getMonth();
    int day = date.getDay();
    if (day < getTotalDayOfMonth(year, month)) {
        day++;
    } else if (month < 12) {
        month++;
        day = 1;
    } else {
        year++;
        month = 1;
        day = 1;
    }
    date.setYear(year);
    date.setMonth(month);
    date.setDay(day);
}

File: 数据库/db_project/src/utils/DateUtil.java
package utils;

import java.text.DateFormat;
import java.text.DecimalFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

public class DateUtil {
    private static DecimalFormat df = new DecimalFormat("0.00");

    public static String dateToString(Date date) {
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        return simpleDateFormat.format(date);
    }

    public static String dateToFirstDayMonth(Date date) {
        return yearMonthToFirstDayMonth(date.getYear() + 1900, date.getMonth() + 1);
    }

    public static String yearMonthToLastDayMonth(int year, int month) {
        int tempYear = year;
        int tempMonth = month + 1;
        if (tempMonth > 12) {
            tempYear++;
            tempMonth -= 12;
        }
        return yearMonthToFirstDayMonth(tempYear, tempMonth);
    }

    public static String yearMonthToFirstDayMonth(int year, int month) {
        return month < 10 ? year + "-0" + month + "-01 00:00:00" : year + "-" + month + "-01 00:00:00";
    }

    public static String yearMonthDayToString(int year, int month, int day) {
        String monthStr = month < 10 ? "0" + month : "" + month;
        String dayStr = day < 10 ? "0" + day : "" + day;
        return year + "-" + month + "-" + day + " 00:00:00";
    }

    public static String dateAdd(Date date, int year, int month) {
        int tempYear = date.getYear() + year + 1900;
        int tempMonth = date.getMonth() + 1 + month;
        int tempDay = date.getDay();
        if (month > 12) {
            tempYear++;
            tempMonth -= 12;
        }
        if (month == 2 && tempDay > 28) {
            tempDay = 28;
        }
        return yearMonthDayToString(tempYear, tempMonth, tempDay);
    }

    public static String getStringByLength(String string, int length) {
        int len = string.length();
        for (int i = len; i < length; i++) {
            string += " ";
        }
        return string;
    }

    public static String formatDouble(double a) {
        return df.format(a);
    }
}
